cpp/src/arrow/compute/row/row_test.cc

```diff
@@ -0,0 +1,121 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <numeric>
+
+#include "arrow/compute/row/encode_internal.h"
+#include "arrow/compute/row/row_internal.h"
+#include "arrow/testing/generator.h"
+#include "arrow/testing/gtest_util.h"
+
+namespace arrow {
+namespace compute {
+
+namespace {
+
+Result<RowTableImpl> MakeRowTableFromColumn(const std::shared_ptr<Array>& column,
+                                            int64_t num_rows, int row_alignment,
+                                            int string_alignment) {
+  DCHECK_GE(column->length(), num_rows);
+  MemoryPool* pool = default_memory_pool();
+
+  std::vector<KeyColumnArray> column_arrays;
+  std::vector<Datum> values{column};
+  ExecBatch batch(std::move(values), num_rows);
+  RETURN_NOT_OK(ColumnArraysFromExecBatch(batch, &column_arrays));
+
+  std::vector<KeyColumnMetadata> column_metadatas;
+  RETURN_NOT_OK(ColumnMetadatasFromExecBatch(batch, &column_metadatas));
+  RowTableMetadata table_metadata;
+  table_metadata.FromColumnMetadataVector(column_metadatas, row_alignment,
+                                          string_alignment);
+
+  RowTableImpl row_table;
+  RETURN_NOT_OK(row_table.Init(pool, table_metadata));
+
+  RowTableEncoder row_encoder;
+  row_encoder.Init(column_metadatas, row_alignment, string_alignment);
+  row_encoder.PrepareEncodeSelected(0, num_rows, column_arrays);
+
+  std::vector<uint16_t> row_ids(num_rows);
+  std::iota(row_ids.begin(), row_ids.end(), 0);
+
+  RETURN_NOT_OK(row_encoder.EncodeSelected(&row_table, static_cast<uint32_t>(num_rows),
+                                           row_ids.data()));
+
+  return std::move(row_table);
+}
+
+}  // namespace
+
+TEST(RowTableMemoryConsumption, Encode) {
+  constexpr int64_t num_rows_max = 8192;
+  constexpr int64_t padding_for_vectors = 64;
```

Comment:
Refers to this https://github.com/apache/arrow/blob/b044a51f410e457592394e25324a5ce6f2276c9a/cpp/src/arrow/compute/row/row_internal.h#L227
which is always appended to the needed capacity when resizing buffers.